### ISSUE TITLE
Reduce querier parquet object store pull latency with async fetch

### ISSUE THREAD

**Issue by philjb** [influxdata/influxdb_iox#5314]
## Background Analogy

TSM has an issue when bringing in new tsm files after a compaction. The compaction has taken several input TSM files and created fewer, new output tsm files. The input files had been in use and because TSM uses mmap as a cache, are likely to be resident in memory if the files are regularly used. The newly compacted files are likely to be larger the same data but because they are different files, the kernel will flush out the removed files and will need to load in the new file's data. 

This has the potential to increase query latency for queries requesting data in the compaction set. 

This latency issue can be solved by prewarming the mmap cache before replacing the input tsm file set. 

## IOx implication

The IOx situation is largely the same. It doesn't use mmap but combines the function of the local disk storage and mmap that TSM uses into just using the object store. 

Newly compacted files are placed in object store and then iox queries, after becoming aware of them, need to fetch them and integrate them so they can be queried.

IOx should plan to asynchronously fetch new compacted files when possible so that the fetching and index reading can occur without queries waiting on it. Queries would continue to go against the parquet file set the querier already has. There is no correctness issue here since compaction reifies deletes and makes the parquet files more performant. The existing set of parquet files has the same data. 

Caveat: async fetch doesn't work, naturally, when a querier needs a parquet file it doesn't have to answer a query. No choice but to stall for the fetch. This is analogous in TSM when the part of a tsm file for a query isn't resident in mmap memory and needs to be fetched from disk. The TSM/cloud2 volume fetch though is likely faster than an object store fetch. 

The internal tsm issue will be back linked.

**Comment by crepererum** [comment]
I'm a bit uncertain about the scope of this issue. Is this about a generic cache priming / warm-up or only about "transfering" the _loaded_ state from an uncompacted to a compacted file? I.e. if we have a file in-cache, it is then compacted, we refresh our "view" about which files exists, then we shall load the compacted file? Now what are the rules for that, because a compacted file may "contain" cached and uncached files?

**Comment by crepererum** [comment]
Also, I would exclude some compactor->querier notification channel or refresh mechanism from this issue and treat that as a separate problem.

**Comment by philjb** [comment]
Goal (via @pauldix): the user has a good experience querying this data that is being moved around by the ingester and the compactor. Such that they don't have to wait for object store requests every time after either of these scenarios:

- Ingester persists new file (particularly for recent time ranges)
- Compactor creates optimized files to remove overlap from L0 or to combine old L1 files into a larger L1 file

**Comment by philjb** [comment]
@crepererum I think Paul summarized the concept (repeated above) that i wanted to capture with this ticket. I included the (tsm) history at how a potentially poor situation could arise with fetching new compacted data.

My technical knowledge of parquet and what the querier does isn't enough for me to evaluate "scope". 
For instance, I don't understand this:
> compacted file may "contain" cached and uncached files?

Please scope this as you see best!

**Comment by crepererum** [comment]
Downgrading prio to "nice to have" and unassigning myself for now, since there are other things that I need to do first and we can technically go live without this feature.

**Comment by crepererum** [comment]
Added https://github.com/influxdata/influxdb_iox/issues/5318 as a first step (doesn't solve the issue, but fixes a related latency problem that is at least as important).
